File: java/src/main/java/br/com/emendes/problems/ScrambleString.java
package br.com.emendes.problems;

import java.util.HashMap;
import java.util.Map;

/**
 * Podemos embaralhar uma string s para obter uma string t usando o seguinte algoritmo:
 * <br>
 * <ol>
 * <li>Se o comprimento da string for 1, pare.</li>
 * <li>Se o comprimento da string for > 1, faça o seguinte:</li>
 *      <ul>
 *        <li>
 *          Divida a string em duas substrings não vazias em um índice aleatório, ou seja,
 *          se a string for s, divida-a em x e y, onde s = x + y.
 *        </li>
 *        <li>
 *          Decida aleatoriamente trocar as duas substrings ou mantê-las na mesma ordem. ou seja, após esta etapa,
 *        s pode se tornar s = x + y ou s = y + x
 *        </li>
 *        <li>
 *          Aplique a etapa 1 recursivamente em cada uma das duas substrings x e y.
 *        </li>
 *      </ul>
 * </ol>
 * <p>
 * Dadas duas strings s1 e s2 do mesmo comprimento, retorne verdadeiro se s2 for uma string embaralhada de s1,
 * caso contrário, retorne falso.<br>
 * <br>
 * Restrições:<br>
 * s1.length == s2.length<br>
 * 1 <= s1.length <= 30<br>
 * s1 e s2 consistem de letras inglêsas minúsculas.
 */
public class ScrambleString {

  private Map<String, Boolean> memo = new HashMap<>();

  public boolean isScramble(String s1, String s2) {
    if (s1.equals(s2)) return true;

    String hashKey = s1 + "+" + s2;
    if (memo.containsKey(hashKey))
      return memo.get(hashKey);

    int[] count = new int[128];

    for (int i = 0; i < s1.length(); i++) {
      count[s1.charAt(i)]++;
      count[s2.charAt(i)]--;
    }

    for (int freq : count) {
      if (freq != 0) {
        memo.put(hashKey, false);
        return false;
      }
    }

    for (int i = 1; i < s1.length(); i++) {
      if (isScramble(s1.substring(0, i), s2.substring(0, i)) &&
          isScramble(s1.substring(i), s2.substring(i))) {
        memo.put(hashKey, true);
        return true;
      }
      if (isScramble(s1.substring(0, i), s2.substring(s2.length() - i)) &&
          isScramble(s1.substring(i), s2.substring(0, s2.length() - i))) {
        memo.put(hashKey, true);
        return true;
      }

    }
    memo.put(hashKey, false);
    return false;
  }

}

File: javascript/problems/ScrambleString.js
/**
 * Nós podemos scramble uma string s para ter uma string t usando o seguinte algoritmo:
 *
 * 1. Se o comprimento da string for 1, pare.
 *
 * 2. Se o comprimento da string for > 1, faça o seguinte:
 * - Divida a string em duas substrings não vazias em um índice aleatório. (i.e. se a string for s,
 * divida em x e y de modo que s = x + y)
 * - Aleatóriamente decida se troca as substrings de ordem ou se mantém a mesma ordem.
 * - Aplique o passo 1 recursivamente em cada substring x e y.
 *
 * Dado duas strings s1 e s2 de mesmo comprimento, retorne true se s2 é uma scrambled string de s1,
 * e false caso contrário.
 *
 * Restrições:
 *
 * s1.length == s2.length
 *
 * 1 <= s1.length <= 30
 *
 * s1 e s2 consistem de letras minúsculas.
 */
const isScramble = function (s1, s2) {
  if (s1 === s2) return true;
  const key = s1 + "+" + s2;
  if (cache[key] !== undefined) return cache[key];
  if (hasDifferentLetters(s1, s2)) return false;

  for (let i = 1; i < s1.length; i++) {
    const s1Left = s1.substring(0, i);
    const s1Right = s1.substring(i);
    const result =
      (isScramble(s1Left, s2.substring(0, i)) &&
        isScramble(s1Right, s2.substring(i))) ||
      (isScramble(s1Left, s2.substring(s2.length - i)) &&
        isScramble(s1Right, s2.substring(0, s2.length - i)));
    if (result) {
      cache[key] = true;
      return true;
    }
  }
  cache[key] = false;
  return false;
};

const cache = {};

const hasDifferentLetters = (s1, s2) => {
  const lettersCount = new Array(26).fill(0);
  for (let i = 0; i < s1.length; i++) {
    lettersCount[s1.charCodeAt(i) - 97]++;
    lettersCount[s2.charCodeAt(i) - 97]--;
  }
  for (const v of lettersCount) {
    if (v < 0) return true;
  }
  return false;
};
